### ISSUE TITLE
API post /api/v0/devices returns all devices

### ISSUE THREAD

**Issue by xantihermo** [librenms/librenms#13822]
### The problem

Hi,

I assume the issue was introduced with the upgrade to 22.1.0-75-g700c6ae9c, always worked before.
When I post to the device endpoint it returns all devices and doesn't create one.
I also see an error with the patch to this endpoint, also worked before:
"message": "Device field to patch has not been supplied"\n}'

Version	22.1.0-75-g700c6ae9c - Tue Feb 15 2022 22:39:23 GMT+0100
Database Schema	2021_12_02_113537_ports_stp_designated_cost_change_to_int (234)
Web Server	nginx/1.18.0
PHP	7.4.3
Python	3.8.10
MySQL	10.3.34-MariaDB-0ubuntu0.20.04.1
Laravel	8.80.0
RRDtool	1.7.2

Regards,

Xanti

### Output of ./validate.php

```txt
librenms@monitoring:~$  /opt/librenms/validate.php
====================================
Component | Version
--------- | -------
LibreNMS  | 22.2.1-10-g86a504e67
DB Schema | 2022_02_03_164059_increase_auth_id_length (235)
PHP       | 7.4.3
Python    | 3.8.10
MySQL     | 10.3.34-MariaDB-0ubuntu0.20.04.1
RRDTool   | 1.7.2
SNMP      | 5.8
====================================

[OK]    Composer Version: 2.2.6
[OK]    Dependencies up-to-date.
[OK]    Database connection successful
[OK]    Database schema correct
librenms@monitoring:~$
```

**Comment by murrant** [comment]
The METHOD of the http request matters.  "Device field to patch has not been supplied" means you have entered the device update path which uses PATCH.  To add a device, you want the POST method.

**Comment by xantihermo** [comment]
Hi,

I was probably not clear with my question or remark, I'm not looking for support but I mentioned 2 issues with the device API endpoint. 

1. A post request gives back all devices (like a get request) and doesn't creates a device, it worked before the mentioned upgrade.
2. A patch request is also broken or changed default behaviour.

Regards,

Xanti

**Comment by murrant** [comment]
Please show your requests.  No one could help if we can't reproduce your issue.

**Comment by xantihermo** [comment]
Hi,

class LibreNmsApi:
    def __init__(self, key, url='http://librenms.jarviss.cloud/api/v0/'):
        self.key = key
        self.url = url
        self.session = requests.Session()
        self.session.verify = False

    def create_devices(self, data=None):
        headers = {'X-Auth-Token': self.key}
        r = self.session.post(self.url+"devices", headers=headers, data=json.dumps(data))
        if r.status_code == 200:
            return r.json()
        return r.json()

    def update_device(self, hostname= "", data=None):
        headers = {'X-Auth-Token': self.key}
        r = self.session.patch(self.url+"devices/"+ hostname, headers=headers, data=json.dumps(data))
        if r.status_code == 200:
            return r.json()
        return r.json()

    def create_parents(self, hostname="", data=None):
        headers = {'X-Auth-Token': self.key}
        r = self.session.post(self.url+"devices/"+ hostname +"/parents", headers=headers, data=json.dumps(data))
        if r.status_code == 201:
            return r.json()
        return r.json()

conn = LibreNmsApi(key="somekey")

data = {
	"hostname": hostname,
	"overwrite_ip": overwrite_ip,
	"version": version,
	"community": community
		
}

output = conn.create_devices(data=data)

--> this provide a list of all configured devices


data = {
	"field": ["notes","purpose"],
	"data":  [ notes, description ]
		
}

output2 = conn.update_device(hostname=hostname, data=data)

--> this gives the error : "message": "Device field to patch has not been supplied"\n}'

Regards,

Xanti

**Comment by murrant** [comment]
Yes, that is what the display field is for. So you can set a name that will be displayed.   overwrite_ip is deprecated, which is why I suggested not using it.

I manually set the input within the python file, here is my complete script (I used force add so I didn't have to put in actual data):
```python
import requests
import json

class LibreNmsApi:
    def __init__(self, key, url='http://librenms.local/api/v0/'):
        self.key = key
        self.url = url
        self.session = requests.Session()
        self.session.verify = False

    def create_devices(self, data=None):
        headers = {'X-Auth-Token': self.key}
        r = self.session.post(self.url+"devices", headers=headers, data=json.dumps(data))
        if r.status_code == 200:
            return r.json()
        return r.json()

    def update_device(self, hostname= "", data=None):
        headers = {'X-Auth-Token': self.key}
        r = self.session.patch(self.url+"devices/"+ hostname, headers=headers, data=json.dumps(data))
        if r.status_code == 200:
            return r.json()
        return r.json()

    def create_parents(self, hostname="", data=None):
        headers = {'X-Auth-Token': self.key}
        r = self.session.post(self.url+"devices/"+ hostname +"/parents", headers=headers, data=json.dumps(data))
        if r.status_code == 201:
            return r.json()
        return r.json()

conn = LibreNmsApi(key="9a0da3ac80c296e97c7db33aba1ab5c0")

data = {
	"hostname": "hostname",
	"overwrite_ip": "overwrite_ip",
	"version": "version",
	"community": "community",
        "version": "v2c",		
        "force_add": True
}

output = conn.create_devices(data=data)

print(output)

data = {
	"field": ["notes","purpose"],
	"data":  [ "notes", "description" ]
		
}

output2 = conn.update_device(hostname="hostname", data=data)

print(output2)
```

**Comment by murrant** [comment]
FYI, for using display instead of overwrite IP, it would look like this:
data = {
	"hostname": "192.168.1.1",
	"display": "Main Router",
	"version": "version",
	"community": "community",
        "version": "v2c"
}

**Comment by xantihermo** [comment]
Hi,

Thanks for the heads up for the deprecated overwrite_ip field, I need to convert all entries and scripts using this field.
I found the problem for the API giving back all devices instead of creating it with a post. Sometime ago I converted the management to https with http redirection, I was still posting to the http url and somehow that caused the issue, if I use https in the requests the API just works. thanks for your time on this issue and it was to my regret a support issue.

Regards,

Xanti
